### PR TITLE
Upgrade baselet

### PR DESCRIPTION
**Pull request by samholmes**
This upgrades baselet to the new branch which includes type safety on all types of baselets. Includes necessary changes to the processor and engine with the added type safety/coverage.

Replaces: https://github.com/EdgeApp/edge-currency-plugins/pull/181

### CODE REVIEW

**Review by TheCharlatan — APPROVED** [review]

File: src/common/utxobased/db/makeProcessor.ts [hunk 1]
```diff
@@ -78,7 +71,7 @@ export interface Processor {
   removeUtxos: (utxoIds: string[]) => Promise<void>
   // fetch either all UTXOs if the array is empty or as selected from an array
   // of UTXO ids
-  fetchUtxos: (args: FetchUtxosArgs) => Promise<IUTXO[]>
+  fetchUtxos: (args: FetchUtxosArgs) => Promise<Array<IUTXO | undefined>>
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:74]: How can you have "undefined"?

File: src/common/utxobased/db/makeProcessor.ts [hunk 2]
```diff
@@ -100,7 +93,7 @@ export interface Processor {
   removeTransaction: (txId: string) => Promise<void>
   fetchTransactions: (
     args: FetchTransactionArgs
-  ) => Promise<IProcessorTransaction[]>
+  ) => Promise<Array<IProcessorTransaction | undefined>>
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:96]: How can you have "undefined"?

File: src/common/utxobased/db/makeProcessor.ts [hunk 3]
```diff
@@ -134,7 +127,7 @@ export interface Processor {
   // insert a new block height / block hash pair. Evicts pairs further back in
   // history than the threshold blocks
   saveBlockHash: (args: BlockHeightArgs) => Promise<void>
-  fetchBlockHash: (height: number) => Promise<string[]>
+  fetchBlockHash: (height: number) => Promise<Array<string | undefined>>
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:130]: How can you have "undefined"?

File: src/common/utxobased/db/makeProcessor.ts [hunk 4]
```diff
@@ -193,47 +186,83 @@ export async function makeProcessor(
...
-            return [undefined]
+            return []
           }
         }
 
         // Return all UTXOs if no UTXO ids are specified
         if (utxoIds.length === 0) {
-          const dump = await tables.utxoById.dumpData('')
-          return Object.values(dump.data)
+          const { data } = await tables.utxoById.dumpData('')
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:264]: Do not use the "dumpData" function for anything other than dumping data.
You should not expect that the return types from this function to not change.

File: src/common/utxobased/db/makeProcessor.ts [hunk 5]
```diff
@@ -320,17 +347,14 @@ export async function makeProcessor(
 
     async fetchTransactions(
       args: FetchTransactionArgs
-    ): Promise<IProcessorTransaction[]> {
+    ): Promise<Array<IProcessorTransaction | undefined>> {
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:350]: How can you have "undefined"?

File: src/common/utxobased/db/makeProcessor.ts [hunk 6]
```diff
@@ -320,17 +347,14 @@ export async function makeProcessor(
...
       const { blockHeightMax, txId, options } = args
       let { blockHeight } = args
-      const txs: IProcessorTransaction[] = []
+      const txs: Array<IProcessorTransaction | undefined> = []
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:353]: How can you have "undefined"?

File: src/common/utxobased/db/makeProcessor.ts [hunk 7]
```diff
@@ -564,7 +583,7 @@ export async function makeProcessor(
       })
     },
 
-    async fetchBlockHash(height: number): Promise<string[]> {
+    async fetchBlockHash(height: number): Promise<Array<string | undefined>> {
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:586]: How can you have "undefined"?

**Review by thehobbit85 — CHANGES_REQUESTED** [review]

File: src/common/utxobased/db/makeProcessor.ts (hunk 1 above)
Comment (samholmes) [reply src/common/utxobased/db/makeProcessor.ts:74]: If the query parameters include a UTXO ID that doesn't exist in the database, then undefined is returned for that entry:

```js
fetchUtxo({ utxoIds: ['x', 'y'] }) // Returns [x, undefined] if 'y' ID does not exist.
```

File: src/common/utxobased/db/makeProcessor.ts (hunk 6 above)
Comment (samholmes) [reply src/common/utxobased/db/makeProcessor.ts:353]: See above.

File: src/common/utxobased/db/makeProcessor.ts (hunk 7 above)
Comment (samholmes) [reply src/common/utxobased/db/makeProcessor.ts:586]: See above.

**Comment by samholmes** [comment]
Approving upon in-person discussion

**Pull request merged**